File: src/main/java/ru/citeck/ecos/predicate/comparator/DefaultValueComparator.java
package ru.citeck.ecos.predicate.comparator;

import java.time.Instant;
import java.time.OffsetDateTime;
import java.time.format.DateTimeParseException;
import java.util.Collection;
import java.util.Date;

public class DefaultValueComparator implements ValueComparator {

    private static final double DOUBLE_THRESHOLD = 0.00000001;

    @Override
    public boolean isEquals(Object value0, Object value1) {

        if (value0 == value1) {
            return true;
        }
        if (value0 == null || value1 == null) {
            return false;
        }
        if (value0.getClass().equals(value1.getClass())) {
            return value0.equals(value1);
        }
        if (value0 instanceof Number && value1 instanceof Number) {
            double v0 = ((Number) value0).doubleValue();
            double v1 = ((Number) value1).doubleValue();
            return Math.abs(v0 - v1) < DOUBLE_THRESHOLD;
        }
        if (value0 instanceof String || value1 instanceof String) {
            return String.valueOf(value0).equals(String.valueOf(value1));
        }

        return false;
    }

    @Override
    public boolean isContains(Object value, Object subValue) {

        if (value == null || subValue == null) {
            return false;
        }

        if (value instanceof Collection) {
            return ((Collection) value).contains(subValue);
        }
        if (value instanceof String || subValue instanceof String) {
            String v0 = String.valueOf(value).toLowerCase();
            String v1 = String.valueOf(subValue).toLowerCase();
            return v0.contains(v1);
        }

        return false;
    }

    @Override
    public boolean isIn(Object value, Object inValue) {

        if (value == null || inValue == null) {
            return false;
        }

        if (inValue instanceof Collection) {
            for (Object collectionValue : (Collection) inValue) {
                if (isEquals(collectionValue, value)) {
                    return true;
                }
            }
        }

        return false;
    }

    @Override
    public boolean isGreaterThan(Object value, Object thanValue, boolean inclusive) {
        return compareGL(value, thanValue, true, inclusive);
    }

    @Override
    public boolean isLessThan(Object value, Object thanValue, boolean inclusive) {
        return compareGL(value, thanValue, false, inclusive);
    }

    @Override
    public boolean isLike(Object value, Object likeValue) {

        if (value == null || likeValue == null) {
            return false;
        }

        String valueStr = String.valueOf(value).toLowerCase();
        String likeStr = String.valueOf(likeValue).toLowerCase();

        likeStr = likeStr.replace("%", ".*");
        likeStr = likeStr.replace("_", ".");

        return valueStr.matches(likeStr);
    }

    @Override
    public boolean isEmpty(Object value) {

        if (value == null) {
            return true;
        }
        if (value instanceof String && ((String) value).isEmpty()) {
            return true;
        }

        return value instanceof Collection && ((Collection) value).isEmpty();
    }

    private boolean compareGL(Object value0,
                              Object value1,
                              boolean isGreater,
                              boolean inclusive) {

        if (value0 == null || value1 == null) {
            return false;
        }

        CompareResult result = compareDouble(value0, value1, isGreater, inclusive);
        if (result == CompareResult.UNKNOWN) {
            result = compareDate(value0, value1, isGreater, inclusive);
        }

        return result == CompareResult.TRUE;
    }

    private CompareResult compareDate(Object value0,
                                Object value1,
                                boolean isGreater,
                                boolean inclusive) {

        Date dateValue0 = parseDate(value0);
        if (dateValue0 == null) {
            return CompareResult.UNKNOWN;
        }
        Date dateValue1 = parseDate(value1);
        if (dateValue1 == null) {
            return CompareResult.UNKNOWN;
        }

        if (isGreater ? dateValue0.after(dateValue1) : dateValue0.before(dateValue1)) {
            return CompareResult.TRUE;
        } else if (inclusive) {
            if (dateValue0.getTime() == dateValue1.getTime()) {
                return CompareResult.TRUE;
            } else {
                return CompareResult.FALSE;
            }
        }

        return CompareResult.FALSE;
    }

    private Date parseDate(Object value) {
        if (value instanceof Date) {
            return (Date) value;
        }
        if (value instanceof String) {
            try {
                value = OffsetDateTime.parse((String) value);
            } catch (DateTimeParseException e) {
                return null;
            }
        }
        if (value instanceof OffsetDateTime) {
            value = ((OffsetDateTime) value).toInstant();
        }
        if (value instanceof Instant) {
            value = Date.from((Instant) value);
        }
        if (value instanceof Date) {
            return (Date) value;
        }
        return null;
    }

    private CompareResult compareDouble(Object value0,
                                  Object value1,
                                  boolean isGreater,
                                  boolean inclusive) {

        Double doubleValue0 = parseDouble(value0);
        if (doubleValue0 == null) {
            return CompareResult.UNKNOWN;
        }
        Double doubleValue1 = parseDouble(value1);
        if (doubleValue1 == null) {
            return CompareResult.UNKNOWN;
        }

        if (isGreater ? doubleValue0 > doubleValue1 : doubleValue0 < doubleValue1) {
            return CompareResult.TRUE;
        } else if (inclusive) {
            if (Math.abs(doubleValue0 - doubleValue1) < DOUBLE_THRESHOLD) {
                return CompareResult.TRUE;
            } else {
                return CompareResult.FALSE;
            }
        }
        return CompareResult.FALSE;
    }

    private Double parseDouble(Object value) {

        double doubleValue0;
        if (value instanceof Number) {
            doubleValue0 = ((Number) value).doubleValue();
        } else {
            try {
                doubleValue0 = Double.parseDouble(value.toString());
            } catch (NumberFormatException e) {
                return null;
            }
        }
        return doubleValue0;
    }

    private enum CompareResult {
        TRUE, FALSE, UNKNOWN
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records2/predicate/comparator/DefaultValueComparator.kt
package ru.citeck.ecos.records2.predicate.comparator

import ru.citeck.ecos.commons.data.DataValue
import java.time.Instant
import java.time.format.DateTimeParseException
import kotlin.math.abs

object DefaultValueComparator : ValueComparator {

    private const val DOUBLE_THRESHOLD = 0.0000001

    operator fun invoke(): DefaultValueComparator = this

    override fun isEquals(value0: DataValue, value1: DataValue): Boolean {
        if (value0 === value1) {
            return true
        }
        if (value0.isNull() || value1.isNull()) {
            return false
        }
        if (value0.isNumber() && value1.isNumber()) {
            val v0: Double = value0.doubleValue()
            val v1: Double = value1.doubleValue()
            return abs(v0 - v1) < DOUBLE_THRESHOLD
        }
        when (compareDateTime(value0, value1, CompareType.EQUALS, true)) {
            CompareResult.TRUE -> return true
            CompareResult.FALSE -> return false
            CompareResult.UNKNOWN -> {}
        }
        if (value0.isTextual() && value1.isNumber() || value0.isNumber() && value1.isTextual()) {
            val v0 = toDouble(value0)
            val v1 = toDouble(value1)
            if (!v0.isNaN() && !v1.isNaN()) {
                return abs(v0 - v1) < DOUBLE_THRESHOLD
            }
        }
        if (value0.isObject() && value1.isTextual()) {
            val val1Str = value1.asText()
            return compareObjStrValues(value0) { it == val1Str }
        }
        if (value0.isTextual() || value1.isTextual()) {
            return value0.asText() == value1.asText()
        }
        return value0 == value1
    }

    override fun isContains(value: DataValue, subValue: DataValue): Boolean {
        if (value.isNull() || subValue.isNull()) {
            return false
        }
        if (value.isArray()) {
            return if (subValue.isArray()) {
                value.any { v ->
                    subValue.any { sv ->
                        isEquals(v, sv)
                    }
                }
            } else {
                value.any { isEquals(it, subValue) }
            }
        }
        if (value.isObject() && subValue.isTextual()) {
            val subValStr = subValue.asText().lowercase()
            return compareObjStrValues(value) { it.lowercase().contains(subValStr) }
        }
        if (value.isTextual() || subValue.isTextual()) {
            val v0 = value.asText().lowercase()
            val v1 = subValue.asText().lowercase()
            return v0.contains(v1)
        }
        return false
    }

    override fun isIn(value: DataValue, inValue: DataValue): Boolean {
        if (value.isNull() || inValue.isNull()) {
            return false
        }
        if (inValue.isArray()) {
            return inValue.any { isEquals(it, value) }
        }
        return false
    }

    override fun isGreaterThan(value: DataValue, thanValue: DataValue, inclusive: Boolean): Boolean {
        return compareGL(value, thanValue, true, inclusive)
    }

    override fun isLessThan(value: DataValue, thanValue: DataValue, inclusive: Boolean): Boolean {
        return compareGL(value, thanValue, false, inclusive)
    }

    override fun isLike(value: DataValue, likeValue: DataValue): Boolean {
        if (value.isNull() || likeValue.isNull()) {
            return false
        }
        var likeStr = likeValue.asText().lowercase()
        likeStr = likeStr.replace("%", ".*")
        likeStr = likeStr.replace("_", ".")
        val regex = likeStr.toRegex()

        if (value.isObject()) {
            return compareObjStrValues(value) { it.lowercase().matches(regex) }
        }
        val valueStr = value.asText().lowercase()
        return valueStr.matches(regex)
    }

    override fun isEmpty(value: DataValue): Boolean {
        if (value.isNull()) {
            return true
        }
        return value.isEmpty()
    }

    private fun compareGL(
        value0: DataValue,
        value1: DataValue,
        isGreater: Boolean,
        inclusive: Boolean
    ): Boolean {

        if (value0.isNull() && value1.isNull()) {
            return inclusive
        }
        val compareType = if (isGreater) { CompareType.GREATER } else { CompareType.LESS }
        var result = compareDateTime(value0, value1, compareType, inclusive)
        if (result == CompareResult.UNKNOWN) {
            result = compareDouble(value0, value1, isGreater, inclusive)
        }
        if (result == CompareResult.UNKNOWN) {
            val intRes = value0.asText().compareTo(value1.asText())
            result = if (intRes == 0) {
                if (inclusive) {
                    CompareResult.TRUE
                } else {
                    CompareResult.FALSE
                }
            } else if (intRes > 0) {
                if (isGreater) {
                    CompareResult.TRUE
                } else {
                    CompareResult.FALSE
                }
            } else {
                if (isGreater) {
                    CompareResult.FALSE
                } else {
                    CompareResult.TRUE
                }
            }
        }
        return result == CompareResult.TRUE
    }

    private fun compareDateTime(
        value0: DataValue,
        value1: DataValue,
        type: CompareType,
        inclusive: Boolean
    ): CompareResult {

        val date0 = parseDateTime(value0) ?: return CompareResult.UNKNOWN
        val date1 = parseDateTime(value1) ?: return CompareResult.UNKNOWN

        val compareRes = date0.compareTo(date1)

        if (inclusive && compareRes == 0) {
            return CompareResult.TRUE
        }
        val boolRes = when (type) {
            CompareType.EQUALS -> compareRes == 0
            CompareType.GREATER -> compareRes > 0
            CompareType.LESS -> compareRes < 0
        }
        return if (boolRes) {
            CompareResult.TRUE
        } else {
            CompareResult.FALSE
        }
    }

    private fun parseDateTime(value: DataValue): Instant? {
        if (!value.isTextual()) {
            return null
        }
        val txt = value.asText()
        if (!txt.contains('T') || !txt.endsWith('Z')) {
            return null
        }
        return try {
            Instant.parse(txt)
        } catch (e: DateTimeParseException) {
            return null
        }
    }

    private fun compareDouble(
        value0: DataValue,
        value1: DataValue,
        isGreater: Boolean,
        inclusive: Boolean
    ): CompareResult {

        val doubleValue0 = toDouble(value0)
        val doubleValue1 = toDouble(value1)

        if (doubleValue0.isNaN() || doubleValue1.isNaN()) {
            return CompareResult.UNKNOWN
        }

        return if (abs(doubleValue0 - doubleValue1) < DOUBLE_THRESHOLD) {
            if (inclusive) {
                CompareResult.TRUE
            } else {
                CompareResult.FALSE
            }
        } else {
            if (isGreater) {
                if (doubleValue0 > doubleValue1) {
                    CompareResult.TRUE
                } else {
                    CompareResult.FALSE
                }
            } else {
                if (doubleValue0 < doubleValue1) {
                    CompareResult.TRUE
                } else {
                    CompareResult.FALSE
                }
            }
        }
    }

    private fun toDouble(value: DataValue): Double {

        if (value.isTextual()) {
            val text = value.asText()
            if (text.contains("T") && text.endsWith("Z")) {
                return try {
                    Instant.parse(text).toEpochMilli().toDouble()
                } catch (e: Exception) {
                    Double.NaN
                }
            }
        }
        if (value.isNull()) {
            return Double.MIN_VALUE
        }
        return value.asDouble(Double.NaN)
    }

    private inline fun compareObjStrValues(obj: DataValue, compare: (String) -> Boolean): Boolean {
        val it = obj.fieldNames()
        while (it.hasNext()) {
            val value = obj[it.next()]
            if (value.isTextual() && compare(value.asText())) {
                return true
            }
        }
        return false
    }

    private enum class CompareResult {
        TRUE, FALSE, UNKNOWN
    }

    private enum class CompareType {
        EQUALS, GREATER, LESS
    }
}
